refactor(renderer): LayeredMaterial splitted and simplified

gchoqueux opened the pull request:
Replace PR  #849 

Work has been done to simplify `LayeredMaterial` and its surrounding
parts, especially with the management of textures in shaders.

`LayeredMaterial` has been splitted, and has now a `LayeredMaterialLayer`
property: this object will contain the textures linked to the material,
and uniforms related to each material.

Uniforms from `LayeredMaterial` are no longer accessible as is, and should
be accessible with directly using the uniform name. The material does
not do a lot of things, apart attaching a layer to it and managing
uniforms.

TileMesh has been simplified, removing all accessors in favor of direct
access to its properties. Material operations that were executed in it
have also been removed.

Shaders have been rewritten to reflect change quoted above. In an
attempt to simplify things and make them more comprehensible, TileFS and
TileVS have been splitted in multiple chunks, which are installed
through ShaderChunk.install.

BREAKING CHANGES: the API of TileMesh and LayeredMaterial have changed.
BREAKING CHANGES: LayeredMaterialConstants and RendererConstant have
been removed.

zarov commented on src/Renderer/Shader/Chunk/color_layers_pars_fragment.glsl:
```diff
@@ -0,0 +1,70 @@
+
+float getBorderDistance(vec2 uv) {
+    vec4 p4 = vec4(uv, 1. - uv);
+    vec2 p2 = min(p4.xy, p4.zw);
```
You can do directly `vec2 p2 = min(uv, 1. -uv);`

mbredif replied:
sure !

zarov commented on src/Renderer/MaterialLayer.js:
```diff
@@ -0,0 +1,145 @@
+                }
+            }
+        }
+        return {};
```
No need for this.

zarov commented on src/Renderer/MaterialLayer.js:
```diff
@@ -0,0 +1,145 @@
+    setTextures(textures, pitchs) {
+        this.dispose(false);
+        for (let i = 0, il = textures.length; i < il; ++i) {
+            // TEST if (textures[i]) is needed ?
```
It seems unnecessary indeed, I don't see a case where `textures[i]` is not set, but one way to confirm this would be to write some tests

zarov commented on src/Renderer/MaterialLayer.js:
```diff
@@ -0,0 +1,145 @@
+            get: () => _valueOpacity,
+            set: (value) => {
+                if (_valueOpacity !== value) {
+                    if (value === 0) {
```
This may cause an error: when setting the opacity to 0 and then setting it back to a higher value, the layer is still invisible on screen.

zarov commented on src/Renderer/Shader/Chunk/elevation_pars_vertex.glsl:
```diff
@@ -0,0 +1,39 @@
+    uniform vec4        elevationOffsetScales[NUM_VS_TEXTURES];
+    uniform int         elevationTextureCount;
+
+    highp float decode32(highp vec4 rgba) {
```
No need to specify `highp` ? The only time this shader is used is with `precision_qualifier`.

mbredif replied:
this function is plainly incorrect if vec4 is not highp, so I d rather keep the modifiers here to document it, even if it is redundant.

zarov replied:
That's weird but ok

zarov commented on src/Renderer/Shader/Chunk/color_layers_pars_fragment.glsl:
```diff
@@ -0,0 +1,70 @@
+    if (textureOffset != layer.textureOffset + int(uv.z) || borderDistance < minBorderDistance ) return vec4(0);
+    vec4 color = texture2D(texture, pitUV(uv.xy, offsetScale));
+    if(color.a > 0.0) {
+        if(layer.effect > 2.0) {
```
This part needs a better thinking, those conditions seem quite ugly imho.

mbredif replied:
agreed, but as it is the old code untouched, I propose to improve it in another PR to keep this one reasonnable (!).

zarov replied:
Yes no problem

zarov reviewed (changes requested):
Quick review, I'll take a better look at it next week

zarov commented on src/Layer/Layer.js:
```diff
@@ -53,6 +53,7 @@ class Layer extends THREE.EventDispatcher {
     constructor(id, type, config = {}) {
         super();
 
+        // ??
```
What is that ?

gchoqueux replied:
I create a issue to explain my comment

gchoqueux replied:
see #921

zarov replied:
Thanks, but please remove these comments, or replace them by linking to the issue

gchoqueux replied:
removed

zarov commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -26,67 +25,7 @@ function getSourceExtent(node, extent, targetLevel, source) {
-    }
-}
-
+// TODO Move to XBilParser
```
Todo when ?

gchoqueux replied:
I have a new PR just after this PR

zarov commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -168,24 +115,29 @@ export function updateLayeredMaterialNodeImagery(context, layer, node, parent) {
+            const tileMT = layer.options.tileMatrixSet || extentsDestination[0].crs();
+            nodeLayer = material.addLayer(layer, tileMT);
+
+            // TODO: Sequence must be returned by parent geometry layer
```
Why not do it now ?

gchoqueux replied:
This PR is already big, it would take a new PR. I add a new issue.

gchoqueux replied:
I open #922

gchoqueux committed:
refactor(renderer): LayeredMaterial splitted and simplified

Worid has been fone to simplify LayeredMaterial and its surrounding
parts, especially with the management of textures in shaders.

LayeredMaterial has been splitted, and has now a LayeredMaterialLayer
property: this object will contain the textures linked to the material,
and uniforms related to each material.

Uniforms from LayeredMaterial are no longer accessible as is, and should
be accessible with directly using the uniform name. The material does
not do a lot of things, apart attaching a layer to it and managing
uniforms.

TileMesh has been simplified, removing all accessors in favor of direct
access to its properties. Material operations that were executed in it
have also been removed.

Shaders have been rewritten to reflect change quoted above. In an
attempt to simplify things and make them more comprehensible, TileFS and
TileVS have been splitted in multiple chunks, which are installed
through ShaderChunk.install.

BREAKING CHANGES: the API of TileMesh and LayeredMaterial have changed.
BREAKING CHANGES: LayeredMaterialConstants and RendererConstant have
been removed.

zarov reviewed (approved):
Ok everything seems good to me, I think we shouldn't have missed anything since it's the third time on this code ;)

gchoqueux commented:
@mbredif and @zarov, thank you for all your efforts!
